### ISSUE TITLE
Tracking issue for `duplicate_matcher_binding_names` compatibility lint

### ISSUE THREAD

**Issue by mark-i-m** [rust-lang/rust#57742]
**What is this lint about**

In https://github.com/rust-lang/rust/issues/57593, it was observed that duplicate macro matcher bindings are allowed in declarations, but unusable in invocations:

```rust
macro_rules! foo {
  ($a:expr, $a:expr) => {}
}
```
This is macro compiles fine on its own, but any attempt to use it will result in an error because `$a` is bound twice. This is likely an oversight in the original implementation, but it really feels like a bug, so we want to make this a hard error.

This compatibility lint warns on such a declaration. The end goal is to make this a hard error in a few releases.

**How to fix this warning/error**

Change the binding names to be different.

**Current status**

- [x] PR #57617 introduces the `duplicate_matcher_binding_names` lint as warn-by-default
- [x] PR https://github.com/rust-lang/rust/pull/59394 makes the `duplicate_matcher_binding_names` lint as deny-by-default
- [ ] PR  #59858 makes the `duplicate_matcher_binding_names` a hard error

**Comment by mark-i-m** [comment]
cc @Centril

**Comment by mark-i-m** [comment]
Note: The future compatibility lint has hit stable by now...

**Comment by Centril** [comment]
@mark-i-m I think we have given it enough time; would you like to prepare a PR that turns it into a hard error directly and let's crater + merge that one?

**Comment by mark-i-m** [comment]
@Centril Sure, but should we not make it deny-by-default first? I don't there is a particular hurry on this, right?

**Comment by mark-i-m** [comment]
Note to self: The next step is to merge this branch: https://github.com/rust-lang/rust/compare/master...mark-i-m:dup-matcher-bindings-3
